### PR TITLE
impl: add function to parse compute lro operation info

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #7658 

Compute uses its own operation types in lieu of google::longrunning::Operation and requires some additional computation to populate the methods it uses for polling LRO (Get and Cancel) methods. This function populates a struct with the relevant pieces for Global, GlobalOrganization, Region, and Zone operations to be used during the LRO polling process.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14328)

### CODE REVIEW

File: google/cloud/internal/rest_lro_helpers.cc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+ComputeOperationInfo ParseComputeOperationInfo(std::string const& self_link) {
+  ComputeOperationInfo info;
+  std::vector<std::string> self_link_parts = absl::StrSplit(self_link, '/');
+  for (std::size_t i = 0; i != self_link_parts.size(); ++i) {
+    if (self_link_parts[i] == "projects") {
+      info.project = self_link_parts[++i];
```
Comment (coryan) [inline google/cloud/internal/rest_lro_helpers.cc:28]: How do we know that `self_link_parts[++i]` is valid?

Maybe 

```cc
  for (auto i = self_link_parts.begin(); i != self_link_parts.end(); ++i) {
    if (*i == "projects" && std::next(i) != self_link_parts.end()) {
      info.project = *(++i);
    } else ...
}
```

Though I think this still accepts invalid input, e.g., what happens if the `self_link` is `projects/foo/projects/bar`?

**Review by coryan — CHANGES_REQUESTED** [review]

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 2 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14328)* status: 2 of 6 files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[`google/cloud/internal/rest_lro_helpers.cc` line 28 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14328#gh-1635606347:-O-B8gle4Kqg6_qZs2Jx:b-q1q8g8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cb7c0e436b613e91e18378cfe9114222c811256b/google/cloud/internal/rest_lro_helpers.cc#L28)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I can go with the nuclear option and use `std::regex` and if we feel its too slow we can revisit. Thoughts?
</blockquote></details>

My guess is that regular expressions will make this even worse.  It seems you would need multiple regex, or deeply nested matchers.

I am fine if your parser accepts invalid input (like the `projects/foo/projects/bar` example).  I am not Okay with crashing on invalid inputs.

**Review by coryan — COMMENTED** [review]
> Reviewed 4 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14328)* status: all files reviewed, 1 unresolved discussion (waiting on @scotthart)

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14328)* status: 4 of 6 files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`google/cloud/internal/rest_lro_helpers.cc` line 28 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14328#gh-1635606347:-O-MYTOJCt970Kdmv5hz:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/cb7c0e436b613e91e18378cfe9114222c811256b/google/cloud/internal/rest_lro_helpers.cc#L28)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

My guess is that regular expressions will make this even worse.  It seems you would need multiple regex, or deeply nested matchers.

I am fine if your parser accepts invalid input (like the `projects/foo/projects/bar` example).  I am not Okay with crashing on invalid inputs.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14328)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @scotthart)

**Pull request merged**